frontend/packages/aquascope-editor/src/editor-utils/stepper.tsx

```diff
@@ -0,0 +1,383 @@
+        {content}
+      </span>
+    ) : diff.type === "None" && !diff.value ? (
+      <span key={content} className="perm-diff-none-low">
+        -
+      </span>
+    ) : null
+  );
+
+  return (
+    <tr>
+      {pathCol}
+      <td>{ico}</td>
+      <td>{permSpans}</td>
+    </tr>
+  );
+};
+
+let stepLocation = (step: PermissionsStateStep): number => {
+  return step.location.char_end;
+};
+
+class PermissionStepTableWidget extends WidgetType {
+  pathsHidden: boolean;
+  tableHidden: boolean;
+  line: Line;
+  rowHTML?: string;
+  constructor(
+    readonly view: EditorView,
+    readonly config: StepperConfig,
+    readonly step: PermissionsStateStep,
+    readonly focusedLines: number[],
+    readonly focusedPaths: Map<number, RegExp>
+  ) {
+    super();
+    this.pathsHidden = true;
+    this.tableHidden = false;
+    this.line = view.state.doc.lineAt(stepLocation(step));
+  }
+
+  eq(other: PermissionStepTableWidget): boolean {
+    // Only one table widget can (should) be on a line.
+    return this.line.number === other.line.number;
+  }
+
+  toDOM() {
+    let container = document.createElement("span");
+
+    let doc = this.view.state.doc;
+    let pos = stepLocation(this.step);
+    let maxLineLen = 0;
+
+    for (let line of doc.iterLines()) {
+      maxLineLen = Math.max(maxLineLen, line.length);
+    }
+    let currLine = this.line;
+    let hideTable = !this.focusedLines.includes(currLine.number);
+    this.tableHidden = hideTable;
+
+    let padding = 2 + maxLineLen - currLine.length;
+    let spaces = " " + "_".repeat(padding);
+
+    let arrowOut = "»";
+    let arrowIn = "«";
+
+    let toggleHiddenPaths = (_e: any) => {
+      this.pathsHidden = !this.pathsHidden;
+      let content = container.querySelector<HTMLDivElement>(
+        ".step-table-content"
+      );
+      let dropdown = container.querySelector<HTMLDivElement>(
+        ".step-table-dropdown"
+      );
+      if (content) {
+        content.style.height = this.pathsHidden ? "0px" : "auto";
+      }
+      if (dropdown) {
+        dropdown.innerHTML = this.pathsHidden ? "..." : "";
+      }
+    };
+
+    let toggleWidget = (_e: any) => {
+      this.tableHidden = !this.tableHidden;
+      let contents = container.querySelector<HTMLDivElement>(
+        ".step-widget-contents"
+      );
+      let btn = container.querySelector<HTMLSpanElement>(
+        ".step-table-collapse"
+      );
+
+      if (contents) {
+        contents.style.width = this.tableHidden ? "0px" : "auto";
+      }
+      if (btn) {
+        btn.innerHTML = this.tableHidden ? arrowOut : arrowIn;
+      }
+    };
+    let matchAll = new RegExp("(.*)?");
+    let r = this.focusedPaths.get(currLine.number) ?? matchAll;
+    let [focusedDiffs, hiddenDiffs] = partition(
+      ([path, _]) => !!path.match(r),
+      this.step.state
+    );
+
+    let hiddenDropdown =
+      hiddenDiffs.length > 0 ? (
+        <>
+          <div className="step-table-dropdown">...</div>
```

Comment:
Let's use [ellipsis-h](https://fontawesome.com/v4/icon/ellipsis-h) from FA so it's vertically-centered.